discovery/gossiper.go

```diff
@@ -2351,91 +2349,112 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []n
-	// a goroutine that will keep on trying sending the message to the
-	// remote peer until it succeeds, or the gossiper shuts down. In case
-	// of success, the message will be removed from the database.
-	d.wg.Add(1)
-	go func() {
-		defer d.wg.Done()
-		for {
-			log.Debugf("Sending AnnounceSignatures for channel "+
-				"%v to remote peer %x", msg.ChannelID,
-				remotePeer.SerializeCompressed())
-			err := d.cfg.SendToPeer(remotePeer, msg)
-			if err == nil {
-				// Sending succeeded, we can
-				// continue the flow.
-				break
+	// deleteMsg is a helper closure that will be used to remove the message
+	// from the store once successfully sent to the peer.
+	deleteMsg := func() {
+		err := d.cfg.DB.Update(func(tx *bolt.Tx) error {
+			messageStore := tx.Bucket(messageStoreKey)
+			if messageStore == nil {
+				return errors.New("message store bucket does " +
+					"not exist, but should")
 			}
 
-			log.Errorf("unable to send AnnounceSignatures message "+
-				"to peer(%x): %v. Will retry when online.",
-				remotePeer.SerializeCompressed(), err)
+			return messageStore.Delete(msgKey)
+		})
+		if err != nil {
+			log.Errorf("Unable to remove stray message from "+
+				"store: %v", err)
+		}
+	}
 
-			peerChan := make(chan lnpeer.Peer, 1)
-			d.cfg.NotifyWhenOnline(remotePeer, peerChan)
+	// With the message persisted, we'll now launch a goroutine that will
+	// continue to retry sending the message to the peer every time we
+	// detect they're online.
+	d.wg.Add(1)
+	go d.sendMsg(msg, peerKey, deleteMsg)
 
-			select {
-			case <-peerChan:
-				// Retry sending.
-				log.Infof("Peer %x reconnected. Retry sending"+
-					" AnnounceSignatures.",
-					remotePeer.SerializeCompressed())
+	return nil
+}
 
-			case <-d.quit:
-				log.Infof("Gossiper shutting down, did not send" +
-					" AnnounceSignatures.")
-				return
-			}
+// sendMsg attempts to re-send the message to the peer every time we detect
+// they're online. The onSuccess callback is optional and is triggered once the
+// message has been sent successfully.
+//
+// NOTE: This must be called as a goroutine.
+func (d *AuthenticatedGossiper) sendMsg(msg lnwire.Message,
+	peerKey *btcec.PublicKey, onSuccess func()) {
+
+	defer d.wg.Done()
+
+	msgName := msg.MsgType().String()
+	compressedKey := peerKey.SerializeCompressed()
+	var shortChanID lnwire.ShortChannelID
+
+	switch msg := msg.(type) {
+	case *lnwire.AnnounceSignatures:
+		shortChanID = msg.ShortChannelID
+	case *lnwire.ChannelUpdate:
+		shortChanID = msg.ShortChannelID
+	}
+
+	log.Debugf("Sending %v for channel(%v) to %x", msgName, shortChanID,
+		compressedKey)
+
+	var peer lnpeer.Peer
+	peerChan := make(chan lnpeer.Peer, 1)
+
+	for {
+		d.cfg.NotifyWhenOnline(peerKey, peerChan)
+
+		select {
+		case peer = <-peerChan:
+		case <-d.quit:
+			return
 		}
 
-		log.Infof("Sent channel announcement proof to remote peer: %x",
-			remotePeer.SerializeCompressed())
-	}()
+		err := peer.SendMessage(true, msg)
```

Comment:
The idea here is that we should _continually_ do this for each peer, as we can't be sure that they've actually fully processed this message once we send it. Consider the case of a mobile phone which is roaming between wi-fi and 4g.

Reply:
So ideally we never exit this loop?

Reply:
How would we know when to stop sending in that case?

Reply:
Why not have one loop per peer w/ outstanding messages? Whenever the peer reconnects, it checks the db for messages it needs to send to that peer. If none are left, the loop exits. 

A message is then sent reliably by first writing it to disk, and then sending it once using `SendMessage` since the message will be resent on subsequent connections. Think the only other caveat is that we would need to spin up a reliable sender whenever we detect that no messages for a given peer existed prior to an insert, or on startup if a non-zero number of messages are present. Should play nicely with async on-success callbacks, since no other coordination is required in this loop.

This will also reduce contention around the server's mutex, since we are only registering for to be notified once per peer instead of once per message.

Reply:
Not that we never exit, but each time they connect, we send the our latest update.

Reply:
Great suggestion @cfromknecht! Dunzo.